### DISCUSSION TITLE
[5.x]: strict_variables

### DISCUSSION THREAD

**Discussion by david-duymelinck** [craftcms/cms#14955] [Q&A] [answered]
### Description

Why is strict_variables set in dev mode? 

### Steps to reproduce

1. set CRAFT_DEV_MODE true
2. add an include twig file in another twig file
3. In the include file set a variable which is an alias for an expected variable, {% set a = data.a %}

### Expected behavior

The include should not generate an error. Because dev mode is stricter than production code. It should be the other way around. 

### Actual behavior

the a variable generates an error.

### Craft CMS version

all

### Installed plugins and versions

-

**Comment by brandonkelly** [comment, marked as answer]
Because referencing nonexistent variables is generally considered a bad practice, and most likely a bug. So `strict_variables` being enabled helps identify probable bugs, in dev environments where the error report won’t take your site down.

> It should be the other way around.

Disagree. Generally the live site should be more resilient so bugs don’t impact front-end traffic.

**Reply by david-duymelinck** [reply]
I did some more investigating and It seems it is up to preference to set strict variables.
I think strict variables is too blunt of a measure to prevent non-existing variables. 

The case that sparked this issue is that we use templates that are included that have optional configuration. 
The configuration is checked in the template. So it is intentional that non-existing variables are called.
So in our case the Craft frontend will break as soon as we use those templates. 
I know we could use a default configuration and merge it if configuration is passed to the template. Or use the default filter.
But I don't see the reason to add more code that only errors in dev mode.

So I massaged Craft to leave the dev mode alone except from the strict variables with

       Event::on(
            View::class,
            View::EVENT_AFTER_CREATE_TWIG,
            function (CreateTwigEvent $event) {
                if($event->twig->isStrictVariables()) {
                    $event->twig->disableStrictVariables();
                }
            }
        );

**Reply by brandonkelly** [reply]
> The case that sparked this issue is that we use templates that are included that have optional configuration.

We have similar things in Craft’s templates, and we just set them all to default values at the top, e.g.

```twig
{% set foo = foo ?? true %}
{% set bar = bar ?? false %}
{% set baz = baz ?? null %}
```
